File: backend/app/services/validation.py
"""
Validation Rules Engine — GSTIN, tax math, place-of-supply, confidence scoring.
"""
import re
from dataclasses import dataclass
from typing import List, Optional, Dict, Any

from app.services.gst_engine import GST_RATES
from app.services.indian_states import INDIAN_STATES, STATE_MAP


@dataclass
class ValidationResult:
    rule_id: str
    rule_name: str
    severity: str  # 'critical' | 'warning'
    passed: bool
    message: str
    details: Optional[Dict[str, Any]] = None


# ─────────────────────────────────────────
# GSTIN Validation
# ─────────────────────────────────────────

GSTIN_REGEX = re.compile(r"^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z]{1}[1-9A-Z]{1}Z[0-9A-Z]{1}$")
GSTIN_CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
GSTIN_CHAR_MAP = {c: i for i, c in enumerate(GSTIN_CHARS)}


def validate_gstin_format(gstin: str) -> ValidationResult:
    """Validate GSTIN 15-character format."""
    if not gstin or gstin.strip() == "":
        return ValidationResult("gstin_format", "GSTIN Format", "warning", True,
                                "GSTIN not provided (B2C transaction)")

    valid = bool(GSTIN_REGEX.match(gstin))
    return ValidationResult(
        "gstin_format", "GSTIN Format", "critical", valid,
        "GSTIN format is valid" if valid else f'Invalid GSTIN format: "{gstin}"'
    )


def validate_gstin_state_code(gstin: str) -> ValidationResult:
    """Validate that GSTIN state code (first 2 digits) is a real Indian state."""
    if not gstin or len(gstin) < 2:
        return ValidationResult("gstin_state", "State Code", "warning", True, "GSTIN not provided")

    state_code = gstin[:2]
    state = STATE_MAP.get(state_code)
    return ValidationResult(
        "gstin_state", "State Code", "critical", state is not None,
        f"State: {state['name']}" if state else f"Invalid state code: {state_code}",
        details={"state_code": state_code, "state_name": state["name"]} if state else None,
    )


def validate_gstin_checksum(gstin: str) -> ValidationResult:
    """Validate GSTIN checksum using Luhn Mod 36 algorithm."""
    if not gstin or len(gstin) != 15:
        return ValidationResult("gstin_checksum", "Checksum", "critical", False,
                                "GSTIN must be 15 characters")

    chars = gstin[:14]
    total = 0
    for i, ch in enumerate(chars):
        char_value = GSTIN_CHAR_MAP.get(ch)
        if char_value is None:
            return ValidationResult("gstin_checksum", "Checksum", "critical", False,
                                    f"Invalid character at position {i + 1}")
        factor = 1 if i % 2 == 0 else 2
        product = char_value * factor
        total += product // 36 + product % 36

    check_digit = (36 - (total % 36)) % 36
    expected = GSTIN_CHARS[check_digit]
    actual = gstin[14]

    return ValidationResult(
        "gstin_checksum", "Checksum", "critical", expected == actual,
        "Checksum valid" if expected == actual else f'Checksum mismatch: expected "{expected}", got "{actual}"'
    )


def validate_gstin(gstin: str) -> List[ValidationResult]:
    """Run all GSTIN validations."""
    if not gstin or gstin.strip() == "":
        return [ValidationResult("gstin_format", "GSTIN Format", "warning", True, "No GSTIN (B2C)")]
    return [
        validate_gstin_format(gstin),
        validate_gstin_state_code(gstin),
        validate_gstin_checksum(gstin),
    ]


def extract_state_from_gstin(gstin: str) -> Optional[Dict[str, str]]:
    """Extract state info from a valid GSTIN."""
    if not gstin or len(gstin) < 2:
        return None
    code = gstin[:2]
    state = STATE_MAP.get(code)
    return {"code": code, "name": state["name"]} if state else None


# ─────────────────────────────────────────
# Tax Math Validation
# ─────────────────────────────────────────

TOLERANCE_PAISE = 100  # ₹1 tolerance for rounding


def validate_gst_rate(rate: float) -> ValidationResult:
    """Check if GST rate is a standard slab."""
    valid = rate in GST_RATES
    return ValidationResult(
        "gst_slab", "GST Slab", "critical", valid,
        f"GST rate {rate}% is valid" if valid else f"GST rate {rate}% is not standard. Valid: {GST_RATES}"
    )


def validate_tax_math(
    taxable_value: int, gst_rate: float, is_inter: bool,
    cgst_amount: int, sgst_amount: int, igst_amount: int,
) -> ValidationResult:
    """Validate line item tax math."""
    if is_inter:
        exp_cgst, exp_sgst = 0, 0
        exp_igst = round(taxable_value * gst_rate / 100)
    else:
        exp_cgst = round(taxable_value * gst_rate / 200)
        exp_sgst = round(taxable_value * gst_rate / 200)
        exp_igst = 0

    ok = (
        abs(cgst_amount - exp_cgst) <= TOLERANCE_PAISE and
        abs(sgst_amount - exp_sgst) <= TOLERANCE_PAISE and
        abs(igst_amount - exp_igst) <= TOLERANCE_PAISE
    )
    return ValidationResult(
        "tax_math", "Tax Calculation", "critical", ok,
        "Tax math correct" if ok else f"Tax mismatch: expected CGST={exp_cgst} SGST={exp_sgst} IGST={exp_igst}",
        details={"expected_cgst": exp_cgst, "expected_sgst": exp_sgst, "expected_igst": exp_igst},
    )


def validate_place_of_supply(
    seller_state: str, buyer_state: str, has_cgst: bool, has_sgst: bool, has_igst: bool,
) -> ValidationResult:
    """Validate GST type matches place of supply."""
    is_inter = seller_state != buyer_state

    if is_inter:
        ok = not has_cgst and not has_sgst and has_igst
        return ValidationResult("pos", "Place of Supply", "critical", ok,
                                f"Inter-state ({seller_state}→{buyer_state}): IGST correct" if ok
                                else "Inter-state must use IGST only")
    else:
        ok = has_cgst and has_sgst and not has_igst
        return ValidationResult("pos", "Place of Supply", "critical", ok,
                                f"Intra-state ({seller_state}): CGST+SGST correct" if ok
                                else "Intra-state must use CGST+SGST")


# ─────────────────────────────────────────
# Confidence Scoring
# ─────────────────────────────────────────

def compute_confidence(results: List[ValidationResult]) -> str:
    """Green / Yellow / Red confidence score."""
    critical_fails = [r for r in results if r.severity == "critical" and not r.passed]
    warnings = [r for r in results if r.severity == "warning" and not r.passed]

    if critical_fails:
        return "red"
    if warnings:
        return "yellow"
    return "green"

File: gst-app/src/lib/validation.ts
// ─────────────────────────────────────────────────────────
// Validation Rules Engine
// ─────────────────────────────────────────────────────────

import { GST_RATES } from './gst-engine';
import { INDIAN_STATES } from './indian-states';

export interface ValidationResult {
    ruleId: string;
    ruleName: string;
    severity: 'critical' | 'warning';
    passed: boolean;
    message: string;
    details?: Record<string, unknown>;
}

export type ConfidenceScore = 'green' | 'yellow' | 'red';

export interface ValidationReport {
    results: ValidationResult[];
    confidenceScore: ConfidenceScore;
    passed: boolean; // no critical failures
}

// ─────────────────────────────────────────
// GSTIN Validation
// ─────────────────────────────────────────

const GSTIN_REGEX = /^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z]{1}[1-9A-Z]{1}Z[0-9A-Z]{1}$/;

const GSTIN_CHAR_MAP: Record<string, number> = {};
'0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('').forEach((c, i) => {
    GSTIN_CHAR_MAP[c] = i;
});

/**
 * Validate GSTIN format (15-char alphanumeric)
 */
export function validateGstinFormat(gstin: string): ValidationResult {
    if (!gstin || gstin.trim() === '') {
        return {
            ruleId: 'gstin_format',
            ruleName: 'GSTIN Format',
            severity: 'warning',
            passed: true,
            message: 'GSTIN not provided (B2C transaction)',
        };
    }

    const valid = GSTIN_REGEX.test(gstin);
    return {
        ruleId: 'gstin_format',
        ruleName: 'GSTIN Format',
        severity: 'critical',
        passed: valid,
        message: valid
            ? 'GSTIN format is valid'
            : `Invalid GSTIN format: "${gstin}". Must be 15 characters (e.g., 27AABCU9603R1ZP)`,
    };
}

/**
 * Validate GSTIN state code (first 2 digits)
 */
export function validateGstinStateCode(gstin: string): ValidationResult {
    if (!gstin || gstin.length < 2) {
        return {
            ruleId: 'gstin_state_code',
            ruleName: 'State Code',
            severity: 'warning',
            passed: true,
            message: 'GSTIN not provided',
        };
    }

    const stateCode = gstin.substring(0, 2);
    const state = INDIAN_STATES.find(s => s.code === stateCode);

    return {
        ruleId: 'gstin_state_code',
        ruleName: 'State Code',
        severity: 'critical',
        passed: !!state,
        message: state
            ? `State code ${stateCode} → ${state.name}`
            : `Invalid state code: ${stateCode}`,
        details: state ? { stateCode, stateName: state.name, isUT: state.ut } : undefined,
    };
}

/**
 * Validate GSTIN checksum using Luhn Mod 36 algorithm
 */
export function validateGstinChecksum(gstin: string): ValidationResult {
    if (!gstin || gstin.length !== 15) {
        return {
            ruleId: 'gstin_checksum',
            ruleName: 'GSTIN Checksum',
            severity: 'critical',
            passed: false,
            message: 'GSTIN must be 15 characters for checksum validation',
        };
    }

    const chars = gstin.substring(0, 14);
    let sum = 0;

    for (let i = 0; i < 14; i++) {
        const charValue = GSTIN_CHAR_MAP[chars[i]];
        if (charValue === undefined) {
            return {
                ruleId: 'gstin_checksum',
                ruleName: 'GSTIN Checksum',
                severity: 'critical',
                passed: false,
                message: `Invalid character in GSTIN at position ${i + 1}`,
            };
        }

        const factor = i % 2 === 0 ? 1 : 2;
        const product = charValue * factor;
        sum += Math.floor(product / 36) + (product % 36);
    }

    const checkDigit = (36 - (sum % 36)) % 36;
    const expectedChar = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'[checkDigit];
    const actualChar = gstin[14];

    return {
        ruleId: 'gstin_checksum',
        ruleName: 'GSTIN Checksum',
        severity: 'critical',
        passed: expectedChar === actualChar,
        message: expectedChar === actualChar
            ? 'GSTIN checksum valid'
            : `GSTIN checksum mismatch: expected "${expectedChar}", got "${actualChar}"`,
    };
}

/**
 * Full GSTIN validation (format + state code + checksum)
 */
export function validateGstin(gstin: string): ValidationResult[] {
    if (!gstin || gstin.trim() === '') {
        return [{
            ruleId: 'gstin_format',
            ruleName: 'GSTIN Format',
            severity: 'warning',
            passed: true,
            message: 'GSTIN not provided (B2C transaction)',
        }];
    }

    return [
        validateGstinFormat(gstin),
        validateGstinStateCode(gstin),
        validateGstinChecksum(gstin),
    ];
}

/**
 * Extract state info from a valid GSTIN
 */
export function extractStateFromGstin(gstin: string): { code: string; name: string } | null {
    if (!gstin || gstin.length < 2) return null;
    const code = gstin.substring(0, 2);
    const state = INDIAN_STATES.find(s => s.code === code);
    return state ? { code: state.code, name: state.name } : null;
}

// ─────────────────────────────────────────
// Tax Math Validation
// ─────────────────────────────────────────

const TOLERANCE_PAISE = 100; // ₹1 tolerance for rounding

export interface TaxMathInput {
    taxableValue: number; // paise
    gstRate: number;
    isInterState: boolean;
    cgstAmount: number;
    sgstAmount: number;
    igstAmount: number;
}

/**
 * Validate GST rate is a standard slab
 */
export function validateGstRate(rate: number): ValidationResult {
    const valid = (GST_RATES as readonly number[]).includes(rate);
    return {
        ruleId: 'gst_slab_valid',
        ruleName: 'GST Slab',
        severity: 'critical',
        passed: valid,
        message: valid
            ? `GST rate ${rate}% is valid`
            : `GST rate ${rate}% is not a standard slab. Valid: ${GST_RATES.join(', ')}%`,
    };
}

/**
 * Validate line item tax math
 */
export function validateTaxMath(input: TaxMathInput): ValidationResult {
    let expectedCgst: number, expectedSgst: number, expectedIgst: number;

    if (input.isInterState) {
        expectedCgst = 0;
        expectedSgst = 0;
        expectedIgst = Math.round(input.taxableValue * input.gstRate / 100);
    } else {
        expectedCgst = Math.round(input.taxableValue * input.gstRate / 200);
        expectedSgst = Math.round(input.taxableValue * input.gstRate / 200);
        expectedIgst = 0;
    }

    const cgstOk = Math.abs(input.cgstAmount - expectedCgst) <= TOLERANCE_PAISE;
    const sgstOk = Math.abs(input.sgstAmount - expectedSgst) <= TOLERANCE_PAISE;
    const igstOk = Math.abs(input.igstAmount - expectedIgst) <= TOLERANCE_PAISE;

    const passed = cgstOk && sgstOk && igstOk;

    return {
        ruleId: 'tax_math_line',
        ruleName: 'Tax Calculation',
        severity: 'critical',
        passed,
        message: passed
            ? 'Tax calculation is correct'
            : `Tax math mismatch: Expected CGST=${expectedCgst} SGST=${expectedSgst} IGST=${expectedIgst}`,
        details: { expectedCgst, expectedSgst, expectedIgst },
    };
}

/**
 * Validate invoice totals match sum of line items
 */
export function validateInvoiceTotals(
    items: { taxableValue: number; cgstAmount: number; sgstAmount: number; igstAmount: number; totalAmount: number }[],
    invoiceTotals: { totalTaxableValue: number; totalCgst: number; totalSgst: number; totalIgst: number; totalAmount: number }
): ValidationResult {
    const sumTaxable = items.reduce((s, i) => s + i.taxableValue, 0);
    const sumCgst = items.reduce((s, i) => s + i.cgstAmount, 0);
    const sumSgst = items.reduce((s, i) => s + i.sgstAmount, 0);
    const sumIgst = items.reduce((s, i) => s + i.igstAmount, 0);
    const sumTotal = items.reduce((s, i) => s + i.totalAmount, 0);

    const taxableOk = Math.abs(invoiceTotals.totalTaxableValue - sumTaxable) <= TOLERANCE_PAISE;
    const cgstOk = Math.abs(invoiceTotals.totalCgst - sumCgst) <= TOLERANCE_PAISE;
    const sgstOk = Math.abs(invoiceTotals.totalSgst - sumSgst) <= TOLERANCE_PAISE;
    const igstOk = Math.abs(invoiceTotals.totalIgst - sumIgst) <= TOLERANCE_PAISE;
    const totalOk = Math.abs(invoiceTotals.totalAmount - sumTotal) <= TOLERANCE_PAISE;

    const passed = taxableOk && cgstOk && sgstOk && igstOk && totalOk;

    return {
        ruleId: 'invoice_total_check',
        ruleName: 'Invoice Total',
        severity: 'critical',
        passed,
        message: passed
            ? 'Invoice totals match sum of line items'
            : 'Invoice totals do not match sum of line items',
        details: { sumTaxable, sumCgst, sumSgst, sumIgst, sumTotal },
    };
}

// ─────────────────────────────────────────
// Place of Supply Validation
// ─────────────────────────────────────────

export function validatePlaceOfSupply(
    sellerState: string,
    buyerState: string,
    hasCgst: boolean,
    hasSgst: boolean,
    hasIgst: boolean
): ValidationResult {
    const isInter = sellerState !== buyerState;

    if (isInter) {
        const passed = !hasCgst && !hasSgst && hasIgst;
        return {
            ruleId: 'place_of_supply',
            ruleName: 'Place of Supply',
            severity: 'critical',
            passed,
            message: passed
                ? `Inter-state (${sellerState} → ${buyerState}): IGST applied correctly`
                : `Inter-state transaction should have IGST only, not CGST/SGST`,
        };
    } else {
        const passed = hasCgst && hasSgst && !hasIgst;
        return {
            ruleId: 'place_of_supply',
            ruleName: 'Place of Supply',
            severity: 'critical',
            passed,
            message: passed
                ? `Intra-state (${sellerState}): CGST+SGST applied correctly`
                : `Intra-state transaction should have CGST+SGST, not IGST`,
        };
    }
}

// ─────────────────────────────────────────
// Confidence Scoring
// ─────────────────────────────────────────

export function computeConfidence(results: ValidationResult[]): ConfidenceScore {
    const criticalFails = results.filter(r => r.severity === 'critical' && !r.passed);
    const warnings = results.filter(r => r.severity === 'warning' && !r.passed);

    if (criticalFails.length > 0) return 'red';
    if (warnings.length > 0) return 'yellow';
    return 'green';
}

/**
 * Run full validation suite on an invoice
 */
export function runValidation(input: {
    buyerGstin?: string;
    sellerStateCode: string;
    buyerStateCode: string;
    isInterState: boolean;
    items: TaxMathInput[];
    totals: { totalTaxableValue: number; totalCgst: number; totalSgst: number; totalIgst: number; totalAmount: number };
}): ValidationReport {
    const results: ValidationResult[] = [];

    // GSTIN validation (if provided)
    if (input.buyerGstin) {
        results.push(...validateGstin(input.buyerGstin));
    }

    // GST rate validation for each item
    input.items.forEach((item, idx) => {
        const rateResult = validateGstRate(item.gstRate);
        rateResult.ruleId = `gst_slab_valid_${idx}`;
        results.push(rateResult);

        const mathResult = validateTaxMath(item);
        mathResult.ruleId = `tax_math_line_${idx}`;
        results.push(mathResult);
    });

    // Invoice totals
    results.push(validateInvoiceTotals(
        input.items.map(i => ({
            taxableValue: i.taxableValue,
            cgstAmount: i.cgstAmount,
            sgstAmount: i.sgstAmount,
            igstAmount: i.igstAmount,
            totalAmount: i.taxableValue + i.cgstAmount + i.sgstAmount + i.igstAmount,
        })),
        input.totals
    ));

    // Place of supply
    const hasCgst = input.totals.totalCgst > 0;
    const hasSgst = input.totals.totalSgst > 0;
    const hasIgst = input.totals.totalIgst > 0;

    // Only check place of supply if there's actually some tax
    if (hasCgst || hasSgst || hasIgst) {
        results.push(validatePlaceOfSupply(
            input.sellerStateCode,
            input.buyerStateCode,
            hasCgst,
            hasSgst,
            hasIgst
        ));
    }

    return {
        results,
        confidenceScore: computeConfidence(results),
        passed: results.every(r => r.severity !== 'critical' || r.passed),
    };
}
